File: src/server/importer/importers/commerz-bank-importer.ts
import { parse } from 'csv-parse';
import fs from 'node:fs/promises';

import { BankTransactionBase } from '../../bank-transaction/models/bank-transaction.model';
import { Exception } from '../../core/errors/exception';
import { MapLocalDate } from '../map-local-date.helper';
import { TransformHelper } from '../transform.helper';
import { ParserFactory } from './importer.service';

export class CommerzBankEnImporter implements ParserFactory {
    static DATE_REGEX = /(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})T\d{2}:\d{2}:\d{2}/u;

    fieldsMap = new Map<keyof BankTransactionBase, number>([
        ['movementName', 9],
        ['date', 11],
        ['dateValue', 1],
        ['value', 4],
        ['details', 10],
    ]);
    mapper = new TransformHelper<number>(this.fieldsMap);
    dateParser = new MapLocalDate(this.fieldsMap, 'dd.MM.yyyy');

    // eslint-disable-next-line class-methods-use-this
    splitMessage(_message: string): [string, string?, string?] {
        let message: string = _message;
        let movementName = '';
        if (message.startsWith('Auszahlung')) {
            message = message.slice('Auszahlung'.length); // Auszahlung length
        } else if (message.startsWith('Kartenzahlung')) {
            message = message.slice('Kartenzahlung'.length); // Kartenzahlung length
        }

        const dateMatch = CommerzBankEnImporter.DATE_REGEX.exec(message);
        const endIndex = message.indexOf('End-to-End-Ref');
        if (dateMatch) {
            const { index } = dateMatch;
            movementName = message.slice(0, index);
            const { year, month, day } = dateMatch.groups ?? {};
            const dateInfo = `${day}.${month}.${year}`;
            // eslint-disable-next-line init-declarations
            let details: string | undefined;
            if (movementName.indexOf('/') > 0) {
                details = movementName.slice(movementName.indexOf('/') + 2).trim();
                movementName = movementName.slice(0, movementName.indexOf('/'));
            }
            return [movementName.trim(), details, dateInfo];
        } else if (endIndex !== -1) {
            return [message.slice(0, endIndex).trim()];
        }
        return [message];
    }

    build(row: unknown[]): BankTransactionBase {
        const splitMessage = this.splitMessage(row[3] as string);
        let [, , newDate] = splitMessage;
        const [movementName, details] = splitMessage;

        if (!newDate) {
            newDate = row[0] as string;
        }
        row.push(movementName, details, newDate);
        this.dateParser.transform(row);

        return this.mapper.map(row as Record<number, unknown>);
    }

    async *create(filePath: string): AsyncGenerator<BankTransactionBase> {
        try {
            const fileData = await fs.readFile(filePath);
            const csvData = parse(fileData, { autoParseDate: false, delimiter: ',', fromLine: 2 });

            for await (const row of csvData) {
                yield this.build(row);
            }
        } catch (error) {
            if (typeof error === 'object' && error && 'code' in error && error.code === 'ENOENT') {
                throw new Exception('E10007', 'CommerzBank file not found', { filePath }, error);
            }
            throw error;
        }
    }

    fileRegex = /Umsaetze_KtoNr.*\.CSV/u;

    key = 'commerz-bank/en';
}

File: swift-server/app/importers/importers/commerz-bank.importer.swift
import CSV
import Foundation
import Vapor

class CommerzBankEnImporter: ParserFactory {
	let DATE_REGEX = try! NSRegularExpression(
		pattern: "(?<year>\\d{4})-(?<month>\\d{2})-(?<day>\\d{2})T\\d{2}:\\d{2}:\\d{2}",
		options: [])
	let transformHelper: TransformHelper<[String?]>
	let key: String = "commerz-bank/en"
	let fileRegex: String = "Umsaetze_KtoNr.*\\.CSV"

	init() {
		let fieldsMap = FieldsMap<Int>(
			movementName: 9,
			date: 11,
			dateValue: 1,
			details: 10, value: 4
		)
		self.transformHelper = TransformHelper(fieldsMap, dateFormat: "dd.MM.yyyy")
	}

	func splitMessage(_ msg: String) throws -> (String, String?, String?) {
		var message = msg

		if message.hasPrefix("Auszahlung") {
			message.removeFirst("Auszahlung".count)
		} else if message.hasPrefix("Kartenzahlung") {
			message.removeFirst("Kartenzahlung".count)
		}

		let dateMatch = DATE_REGEX.firstMatch(
			in: message, options: [], range: NSRange(location: 0, length: message.count)
		)
		if let dateMatch = dateMatch {
			let range = Range(dateMatch.range(at: 0), in: message)
			guard let range = range else {
				throw Exception(.E10000)
			}
			var movementName = String(message[..<range.lowerBound])
			let dayRange = Range(dateMatch.range(withName: "day"), in: message)
			let monthRange = Range(dateMatch.range(withName: "month"), in: message)
			let yearRange = Range(dateMatch.range(withName: "year"), in: message)
			let dateInfo =
				"\(message[dayRange!]).\(message[monthRange!]).\(message[yearRange!])"
			var details: String?
			if let slashIndex = movementName.firstIndex(of: "/") {
				details = String(movementName[slashIndex...]).trimmingCharacters(
					in: .whitespaces)
				movementName.removeSubrange(slashIndex...)
			}

			return (
				movementName.trimmingCharacters(in: .whitespaces), details, dateInfo
			)
		} else if let strRange = message.range(of: "End-to-End-Ref") {
			let splitMsg = message[..<strRange.lowerBound]
			return (splitMsg.trimmingCharacters(in: .whitespaces), nil, nil)
		}

		return (message.trimmingCharacters(in: .whitespaces), nil, nil)
	}

	func create(filePath: String) -> AsyncThrowingStream<PartialBankTransaction, Error> {
		AsyncThrowingStream { continuation in
			Task {
				do {
					let csv = try parseCsv(filePath: filePath)

					var lineCounter = 0
					while let row = csv.next() {
						lineCounter += 1
						do {
							var mappedData: [String?] = row

							let (movementName, details, newDate) =
								try splitMessage(
									mappedData[3] ?? "")

							mappedData.append(movementName)
							mappedData.append(details)
							if let newDate = newDate {
								mappedData.append(newDate)
							} else {
								mappedData.append(row[0])
							}
							let transaction = try self.transformHelper
								.map(mappedData)
							continuation.yield(transaction)
						} catch {
							throw Exception(
								.E10011,
								context: ["line": lineCounter],
								cause: error)
						}
					}

					continuation.finish()
				} catch {
					continuation.finish(throwing: error)
				}
			}
		}
	}
}
